### PR TITLE
Check dialect - wip

### PR DESCRIPTION
**Pull request by KavithaTipturMadhu**
@chelini I tried adding a new check dialect for test operations but the dialect doesn't seem to be getting registered. I don't see it in the available dialects list. Can you please take a look to see what I may have missed?

### CODE REVIEW

**Comment by chelini** [comment]
> 

You need to add the dialect here: https://github.com/plaidml/tpp-sandbox/blob/cb2488f06dc57ea8c6eb3dc857b62347d6b5aa2e/standalone-opt/standalone-opt.cpp#L35

File: include/Standalone/Dialect/Check/CheckOps.td [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    any non-zero integer.
+
+    Issues a non-fatal failure if the verification fails.
+
+    ```mlir
+    check.expect_true(%arg0) : i32
+    ```
+  }];
+
+  let arguments = (ins AnySignlessInteger:$operand);
```
Comment (rengolin) [inline include/Standalone/Dialect/Check/CheckOps.td:33]: Do we want to restrict this to booleans (`i1`)?

File: include/Standalone/Dialect/Check/CheckOps.td [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+  let assemblyFormat = "`(` $operand `)` attr-dict `:` type($operand)";
+}
+
+def CHECK_ExpectAlmostEqOp :
+    Op<CHECK_Dialect, "expect_almost_eq", [SameTypeOperands]> {
+  let summary = [{Checks that the operands are almost equal}];
+  let description = [{
+    Verifies that the tensor operands with float elements are
+    almost equal to within an implementation-defined "reasonable" tolerance.
```
Comment (rengolin) [inline include/Standalone/Dialect/Check/CheckOps.td:43]: I think what's _reasonable_ should be a part of this op, via an attribute. We can start with a simple fp-comparison and a delta (which is the attribute), but we could upgrade this into more complex attributes to handle statistical analysis later.

Comment (chelini) [reply include/Standalone/Dialect/Check/CheckOps.td:43]: Agree. I think TOSA has something along this line.

File: include/TPP/Dialect/Check/CheckOps.h [hunk 3]
```diff
@@ -0,0 +1,20 @@
+//===- XsmmOps.h - Xsmm dialect ops -----------------------------*- C++ -*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef CHECK_STANDALONE_OPS_H
```
Comment (chelini) [inline include/TPP/Dialect/Check/CheckOps.h:9]: Do use STANDALONE. We have TPP now

File: include/TPP/Dialect/Check/CheckOps.h [hunk 4]
```diff
@@ -0,0 +1,20 @@
+//===- XsmmOps.h - Xsmm dialect ops -----------------------------*- C++ -*-===//
```
Comment (chelini) [inline include/TPP/Dialect/Check/CheckOps.h:1]: Please update.

File: include/TPP/Dialect/Check/CheckOps.td [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    any non-zero integer.
+
+    Issues a non-fatal failure if the verification fails.
+
+    ```mlir
+    check.expect_true(%arg0) : i1
+    ```
+  }];
+
+  let arguments = (ins AnyI1:$operand);
```
[the remark this answers is not in the document]
Comment (KavithaTipturMadhu) [reply include/TPP/Dialect/Check/CheckOps.td:33]: IIUC, expect_true op needs to check only an I1 unlike the iree version that takes i32. Is this right?

[the remark this answers is not in the document]
Comment (chelini) [reply include/TPP/Dialect/Check/CheckOps.td:33]: Agree. Just wondering what the prefix `Any` means in this context.

File: lib/TPP/ConvertCheckToFunc.cpp [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+using namespace mlir;
+using namespace mlir::check;
+
+#define GEN_PASS_CLASSES
+#include "Standalone/Passes.h.inc"
+#include <iostream>
+namespace {
+
+static func::CallOp buildExpectTrueCall(Location loc, Operation *op,
```
[the remark this answers is not in the document]
Comment (KavithaTipturMadhu) [reply lib/TPP/ConvertCheckToFunc.cpp:26]: you mean add lower calls in xsmm conversion pass? Parts of the code is similar between xsmm and check lowering, but ops are different, and there are no return values from check ops, so things like replaceUsesWith won't work.

[the remark this answers is not in the document]
Comment (chelini) [reply lib/TPP/ConvertCheckToFunc.cpp:26]: no no, here are good. Just wondering if we can share some code, but maybe this is not the case. Just ignore the comment for now

File: include/TPP/Dialect/Check/CheckOps.td (hunk 5 above)
[the remark this answers is not in the document]
Comment (rengolin) [reply include/TPP/Dialect/Check/CheckOps.td:33]: Probably just `i1` would do.

File: include/TPP/Dialect/Check/CheckOps.td [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+  let assemblyFormat = "`(` $operand `)` attr-dict `:` type($operand)";
+}
+
+def CHECK_ExpectAlmostEqOp :
```
Comment (rengolin) [inline include/TPP/Dialect/Check/CheckOps.td:38]: I don't like the word "almost". It doesn't convey what we're trying to do. In the LLVM test suite / benchmarks we use `fp-comp`. We may get into a state that each number isn't even remotely similar to the next, but the overall distribution is, so this is just a special type of comparison.

We'll probably need to compare integers at some point (INT8), so this also can't be called `FPComp`. I don't know a good name, but "almost" isn't it.

File: lib/TPP/Dialect/Check/CheckDialect.cpp [hunk 8]
```diff
@@ -0,0 +1,25 @@
+// Copyright 2020 The IREE Authors
```
Comment (rengolin) [inline lib/TPP/Dialect/Check/CheckDialect.cpp:1]: We cannot attribute copyright to other authors. This line has to go.

Keep them all like the LLVM ones, since we want these files to be in LLVM soon,

File: include/TPP/Dialect/Check/CheckDialect.h [hunk 9]
```diff
@@ -0,0 +1,24 @@
+// Copyright 2020 The IREE Authors
```
[the remark this answers is not in the document]
Comment (rengolin) [reply include/TPP/Dialect/Check/CheckDialect.h:1]: I don't think it's necessary. We can say this in the commit message, in the RFC discussion, etc.

Let's keep those like the LLVM ones, because even if IREE wanted to upstream them, they'd have to drop their copyright comments.

File: include/TPP/Dialect/Check/CheckOps.td (hunk 5 above)
[the remark this answers is not in the document]
Comment (KavithaTipturMadhu) [reply include/TPP/Dialect/Check/CheckOps.td:31]: ```AnyI``` refers to any integer regardless of its signedness semantics, whereas ```I``` refers to signless integer of specific width. I guess using ```I<width>``` makes more sense here.

File: lib/TPP/ConvertCheckToFunc.cpp [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+  auto call = rewriter.create<func::CallOp>(loc, fnName.getValue(), None,
+                                            op->getOperands());
+  return call;
+}
+
+static SmallVector<Value> getMemRefOperands(OpBuilder &b, Location loc,
+                                            ValueRange operands) {
+  SmallVector<Value> res;
+  res.reserve(operands.size());
+  for (Value op : operands) {
```
Comment (chelini) [inline lib/TPP/ConvertCheckToFunc.cpp:55]: I would switch the else with the if here. Write something like:
``` 
if (auto memrefType = op.getType().dyn_cast<MemRefType>()) {
   /* do else stmts*/
}
else {
  /* do if stmts */
}
```

File: lib/TPP/ConvertCheckToFunc.cpp [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+    Location loc = expectTrueOp.getLoc();
+    std::string kindAsString = "expect_true";
+
+    FlatSymbolRefAttr fnName =
+        SymbolRefAttr::get(rewriter.getContext(), kindAsString);
+
+    ModuleOp module = expectTrueOp->getParentOfType<ModuleOp>();
+
+    buildExpectTrueCall(loc, expectTrueOp, module, fnName, rewriter);
+    expectTrueOp.erase();
```
Comment (chelini) [inline lib/TPP/ConvertCheckToFunc.cpp:110]: mm, I think you want to notify the rewriter. `rewriter.eraseOp(exepectedTrueOp)`.

File: lib/TPP/ConvertCheckToFunc.cpp [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+    Location loc = almostEqualsOp.getLoc();
+    std::string kindAsString = "expect_almost_equals";
+
+    FlatSymbolRefAttr fnName =
+        SymbolRefAttr::get(rewriter.getContext(), kindAsString);
+
+    ModuleOp module = almostEqualsOp->getParentOfType<ModuleOp>();
+
+    buildExpectAlmostEqualsCall(loc, almostEqualsOp, module, fnName, rewriter);
+    almostEqualsOp.erase();
```
Comment (chelini) [inline lib/TPP/ConvertCheckToFunc.cpp:130]: same here. Have a look at: https://mlir.llvm.org/docs/PatternRewriter/#restrictions

File: lib/TPP/ConvertCheckToFunc.cpp [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+    return success();
+  }
+};
+
+struct ConvertCheckToFunc : public ConvertCheckToFuncBase<ConvertCheckToFunc> {
+  ConvertCheckToFunc() = default;
+  void runOnOperation() override {
+    RewritePatternSet patterns(&getContext());
+    tpp::populateCheckToFuncPatterns(patterns);
```
Comment (chelini) [inline lib/TPP/ConvertCheckToFunc.cpp:139]: mlir::tpp:: populateCheckToFuncPatterns

File: include/TPP/Dialect/Check/CheckOps.td (hunk 7 above)
Comment (rengolin) [reply include/TPP/Dialect/Check/CheckOps.td:36]: How about `expect_within_range`?

Comment (KavithaTipturMadhu) [reply include/TPP/Dialect/Check/CheckOps.td:36]: That name seems to suggest there would be a range minimum and maximum that are passed as arguments to the func. Shall we decide on the name later and get the PR reviewed for now?

**Comment by KavithaTipturMadhu** [comment]
@chelini I will create new issues for all the changes we have discussed so far with the PR. Shall we review it in its current state and merge for now?

File: include/TPP/Dialect/Check/CheckOps.td [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+    Op<CHECK_Dialect, "expect_almost_eq",[TypesMatchWith<"Operand types match",
+                   "lhs", "rhs",
+                   "$_self.cast<ShapedType>()">]> {
+  let summary = [{Checks that the operands are almost equal}];
+  let description = [{
+    Verifies that the tensor operands with float elements are
+    almost equal to within an implementation-defined "reasonable" tolerance.
+
+    Issues a non-fatal failure if the verification fails.
```
Comment (chelini) [inline include/TPP/Dialect/Check/CheckOps.td:45]: If we assert this is not the case, I believe.

File: lib/TPP/Dialect/Check/BufferizableOpInterfaceImpl.cpp [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+
+  SmallVector<OpResult> getAliasingOpResult(Operation *op, OpOperand &opOperand,
+                                            const AnalysisState &state) const {
+    return {};
+  }
+
+  BufferRelation bufferRelation(Operation *op, OpResult opResult,
+                                const AnalysisState &state) const {
+    return BufferRelation::Equivalent;
+  }
```
Comment (chelini) [inline lib/TPP/Dialect/Check/BufferizableOpInterfaceImpl.cpp:50]: nit add space.

File: lib/TPP/Dialect/Check/BufferizableOpInterfaceImpl.cpp [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
```
Comment (chelini) [inline lib/TPP/Dialect/Check/BufferizableOpInterfaceImpl.cpp:92]: same here.

File: lib/TPP/Dialect/Check/CMakeLists.txt [hunk 17]
```diff
@@ -0,0 +1,23 @@
+add_mlir_dialect_library(TPPCheckDialect
+  # Ops and dialects
```
Comment (chelini) [inline lib/TPP/Dialect/Check/CMakeLists.txt:2]: can you please align the comment with the file names?

File: lib/TPP/Dialect/Check/CheckOps.cpp [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/IR/OpImplementation.h"
+#include "mlir/IR/PatternMatch.h"
+
+namespace mlir {
+namespace check {
+namespace {
+template <typename SrcOp, typename DstOp>
+struct ExpandAttributeToConst : public OpRewritePattern<SrcOp> {
```
Comment (chelini) [inline lib/TPP/Dialect/Check/CheckOps.cpp:12]: Why this pattern?

File: tpp-rt/CheckRunnerUtils.h [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+// retargetable, including on targets without a C++ runtime.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef TPP_EXECUTIONENGINE_CRUNNERUTILS_H
+#define TPP_EXECUTIONENGINE_CRUNNERUTILS_H
+
+#include "mlir/ExecutionEngine/RunnerUtils.h"
+
+// TODO: here we want to have dispatch/invoke only for unary/binary and ternary.
```
Comment (chelini) [inline tpp-rt/CheckRunnerUtils.h:20]: this comment is not related, please remove it.

File: lib/TPP/ConvertCheckToFunc.cpp [hunk 20]
```diff
@@ -0,0 +1,172 @@
...
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "TPP/Dialect/Check/CheckOps.h"
+#include "TPP/Passes.h"
+#include "TPP/Transforms.h"
```
Comment (chelini) [inline lib/TPP/ConvertCheckToFunc.cpp:11]: not needed.

File: lib/TPP/ConvertCheckToFunc.cpp [hunk 21]
```diff
@@ -0,0 +1,172 @@
...
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
```
Comment (chelini) [inline lib/TPP/ConvertCheckToFunc.cpp:14]: not needed

**Review by chelini — APPROVED** [review]
@KavithaTipturMadhu thanks! Looks good but I would like to add the following tests:
1. We need to check that the IR roundtrip properly (`expect_true` is not tested), please add a test case similar to this one: https://github.com/plaidml/tpp-sandbox/blob/main/test/TPP/tpp-ops.mlir
2. Please add a smaller test case that exercise only the lowerings of the ops to the func dialect.

File: lib/TPP/ConvertCheckToFunc.cpp (hunk 20 above)
Comment (KavithaTipturMadhu) [reply lib/TPP/ConvertCheckToFunc.cpp:11]: Method populateCheckToFuncPatterns is defined in TPP/Transforms.h. Do you recommend moving it elsewhere?

Comment (chelini) [reply lib/TPP/ConvertCheckToFunc.cpp:11]: Ah, then it is fine

File: lib/TPP/ConvertCheckToFunc.cpp (hunk 21 above)
Comment (KavithaTipturMadhu) [reply lib/TPP/ConvertCheckToFunc.cpp:14]: I can only remove Arith header from this list. There are dependences on other headers that I cant get rid of.
